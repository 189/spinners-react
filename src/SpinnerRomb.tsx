import React, { CSSProperties } from 'react';
import { defaultProps, normalizeSize, SpinnerProps } from './helpers';

import './SpinnerRomb.css';

const coords = [
  { x: 18.5, y: 33.1901 },
  { x: 18.5, y: 48.1901 },
  { x: 18.5, y: 63.1901 },
  { x: 3.5, y: 48.1901 },
  { x: 33.5, y: 48.1901 },
  { x: 79.5, y: 48.1901 },
  { x: 33.5, y: 33.1901 },
  { x: 33.5, y: 18.1901 },
  { x: 94.5, y: 48.1901 },
  { x: 79.5, y: 33.1901 },
  { x: 33.5, y: 63.1901 },
  { x: 33.5, y: 78.1901 },
  { x: 49.5, y: 18.1901 },
  { x: 49.5, y: 3.19006 },
  { x: 49.5, y: 48.1901 },
  { x: 49.5, y: 33.1901 },
  { x: 49.5, y: 63.1901 },
  { x: 49.5, y: 78.1901 },
  { x: 49.5, y: 93.1901 },
  { x: 64.5, y: 18.1901 },
  { x: 64.5, y: 48.1901 },
  { x: 64.5, y: 33.1901 },
  { x: 64.5, y: 63.1901 },
  { x: 64.5, y: 78.1901 },
  { x: 79.5, y: 63.1901 },
];

export type SpinnerRombProps = SpinnerProps;

export const SpinnerRomb = ({
  color,
  enabled,
  size,
  speed,
  still,
  thickness,
  style,
  ...svgProps
}: SpinnerRombProps) => {
  const svgStyle: CSSProperties = {
    color,
    overflow: 'visible',
    width: normalizeSize(size),
    ...style,
  };
  const rombStyle: CSSProperties = {
    animation: `spinners-react-romb ${140 / speed}s steps(2, end) infinite`,
  };

  if (still) rombStyle.animation = 'none';
  if (!enabled) return null;

  return (
    <svg fill="currentColor" {...svgProps} style={svgStyle} viewBox="0 0 97 96">
      {coords.map((c) => (
        <circle key={`${c.x}-${c.y}`} cx={c.x} cy={c.y} fill="rgba(0,0,0,0.44)" r={2.5 * (thickness / 100)} />
      ))}
      <g style={rombStyle}>
        {coords.filter((c, i) => i < 5).map((c) => (
          <circle key={`h-${c.x}-${c.y}`} cx={c.x} cy={c.y} r={3.5 * (thickness / 100)} />
        ))}
      </g>
    </svg>
  );
};
SpinnerRomb.defaultProps = defaultProps;
