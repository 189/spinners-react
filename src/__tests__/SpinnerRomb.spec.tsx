import React from 'react';
import { create } from 'react-test-renderer';

import { SpinnerRomb } from '../SpinnerRomb';

describe('SpinnerRomb', () => {
  const color = 'red';

  it('matches the snapshot', () => {
    const component = create(<SpinnerRomb color="#fff" size={100} speed={10} thickness={50} />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('matches the snapshot with default props', () => {
    const component = create(<SpinnerRomb />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('renders null if disabled', () => {
    const component = create(<SpinnerRomb enabled={false} />);

    expect(component.toJSON()).toBe(null);
  });

  it('renders still if specified', () => {
    const component = create(<SpinnerRomb still />);
    const group = component.root.findByType('g');

    expect(group.props.style.animation).toBe('none');
  });

  it('passes props to styles', () => {
    const size = '20%';
    const thickness = 40;
    const speed = 50;
    const component = create(
      <SpinnerRomb color={color} size={size} speed={speed} thickness={thickness} />,
    );
    const { props: { style } } = component.toJSON();
    const circles = component.root.findAllByType('circle');
    const last = circles.length - 1;
    const group = component.root.findByType('g');

    expect(style.width).toBe(size);
    expect(circles[0].props.r).toBeCloseTo(2.5 * (thickness / 100));
    expect(circles[last].props.r).toBeCloseTo(3.5 * (thickness / 100));
    expect(group.props.style.animation.includes(140 / speed)).toBe(true);
  });

  it('passes svg props overriding styles', () => {
    const className = 'test-class';
    const component = create(<SpinnerRomb className={className} color="green" style={{ color }} />);
    const { props } = component.toJSON();

    expect(props.className).toBe(className);
    expect(props.style.color).toBe(color);
  });
});
