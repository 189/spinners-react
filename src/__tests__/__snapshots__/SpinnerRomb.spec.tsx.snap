// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpinnerRomb exists for backward compatibility 1`] = `
<svg
  fill="currentColor"
  style={
    Object {
      "color": "#fff",
      "overflow": "visible",
      "width": "100px",
    }
  }
  viewBox="0 0 96 96"
>
  <circle
    cx={3}
    cy={48}
    fill="rgba(0,0,0,0.44)"
    r={1.25}
  />
  <circle
    cx={18}
    cy={33}
    fill="rgba(0,0,0,0.44)"
    r={1.25}
  />
  <circle
    cx={18}
    cy={48}
    fill="rgba(0,0,0,0.44)"
    r={1.25}
  />
  <circle
    cx={18}
    cy={63}
    fill="rgba(0,0,0,0.44)"
    r={1.25}
  />
  <circle
    cx={33}
    cy={48}
    fill="rgba(0,0,0,0.44)"
    r={1.25}
  />
  <circle
    cx={33}
    cy={18}
    fill="rgba(0,0,0,0.44)"
    r={1.25}
  />
  <circle
    cx={33}
    cy={33}
    fill="rgba(0,0,0,0.44)"
    r={1.25}
  />
  <circle
    cx={33}
    cy={63}
    fill="rgba(0,0,0,0.44)"
    r={1.25}
  />
  <circle
    cx={33}
    cy={78}
    fill="rgba(0,0,0,0.44)"
    r={1.25}
  />
  <circle
    cx={49}
    cy={3}
    fill="rgba(0,0,0,0.44)"
    r={1.25}
  />
  <circle
    cx={49}
    cy={18}
    fill="rgba(0,0,0,0.44)"
    r={1.25}
  />
  <circle
    cx={49}
    cy={33}
    fill="rgba(0,0,0,0.44)"
    r={1.25}
  />
  <circle
    cx={49}
    cy={48}
    fill="rgba(0,0,0,0.44)"
    r={1.25}
  />
  <circle
    cx={49}
    cy={63}
    fill="rgba(0,0,0,0.44)"
    r={1.25}
  />
  <circle
    cx={49}
    cy={78}
    fill="rgba(0,0,0,0.44)"
    r={1.25}
  />
  <circle
    cx={49}
    cy={93}
    fill="rgba(0,0,0,0.44)"
    r={1.25}
  />
  <circle
    cx={64}
    cy={18}
    fill="rgba(0,0,0,0.44)"
    r={1.25}
  />
  <circle
    cx={64}
    cy={33}
    fill="rgba(0,0,0,0.44)"
    r={1.25}
  />
  <circle
    cx={64}
    cy={48}
    fill="rgba(0,0,0,0.44)"
    r={1.25}
  />
  <circle
    cx={64}
    cy={63}
    fill="rgba(0,0,0,0.44)"
    r={1.25}
  />
  <circle
    cx={64}
    cy={78}
    fill="rgba(0,0,0,0.44)"
    r={1.25}
  />
  <circle
    cx={79}
    cy={33}
    fill="rgba(0,0,0,0.44)"
    r={1.25}
  />
  <circle
    cx={79}
    cy={48}
    fill="rgba(0,0,0,0.44)"
    r={1.25}
  />
  <circle
    cx={79}
    cy={63}
    fill="rgba(0,0,0,0.44)"
    r={1.25}
  />
  <circle
    cx={94}
    cy={48}
    fill="rgba(0,0,0,0.44)"
    r={1.25}
  />
  <g
    style={
      Object {
        "animation": "spinners-react-diamond 14s steps(2, end) infinite",
      }
    }
  >
    <circle
      cx={3}
      cy={48}
      r={1.75}
    />
    <circle
      cx={18}
      cy={33}
      r={1.75}
    />
    <circle
      cx={18}
      cy={48}
      r={1.75}
    />
    <circle
      cx={18}
      cy={63}
      r={1.75}
    />
    <circle
      cx={33}
      cy={48}
      r={1.75}
    />
  </g>
</svg>
`;
